wodles/azure/tests/test_azure.py

```diff
@@ -0,0 +1,859 @@
+@patch('azure-logs.read_auth_file')
+def test_start_graph(mock_auth, mock_token, mock_build, mock_graph, tenant, auth_path, graph_id, key, offset, query):
+    """Test start_graph attempts to process the logs available for the given authentication, query and offset values."""
+    azure.args = MagicMock(graph_tenant_domain=tenant, graph_auth_path=auth_path, graph_id=graph_id, graph_key=key,
+                           graph_time_offset=offset, graph_query=query)
+    mock_auth.return_value = credentials = ("client", "secret")
+    mock_token.return_value = token = "token"
+    mock_build.return_value = url = "url"
+
+    azure.start_graph()
+
+    if auth_path and tenant:
+        mock_auth.assert_called_with(auth_path=auth_path, fields=("application_id", "application_key"))
+
+    mock_token.assert_called_with(client_id=credentials[0], secret=credentials[1], domain=tenant,
+                                  scope=f"{azure.URL_GRAPH}/.default")
+    md5_hash = md5(query.encode()).hexdigest()
+    mock_build.assert_called_with(offset=offset, md5_hash=md5_hash)
+    mock_graph.assert_called_with(url=url, headers={'Authorization': f'Bearer {token}'}, md5_hash=md5_hash)
+
+
+@patch('azure-logs.logging.error')
+@patch('azure-logs.get_graph_events', side_effect=HTTPError)
+@patch('azure-logs.build_graph_url')
+@patch('azure-logs.get_token')
+@patch('azure-logs.read_auth_file', return_value=("client", "secret"))
+def test_start_graph_ko(mock_auth, mock_token, mock_build, mock_get, mock_logging):
+    """Test start_graph shows error message if get_log_analytics_events returns an HTTP error."""
+    azure.args = MagicMock(graph_id="test", graph_key="test", graph_tenant_domain="test", graph_query="")
+    azure.start_graph()
+    mock_logging.assert_called_once()
+
+
+@patch('azure-logs.logging.error')
+def test_start_graph_ko_credentials(mock_logging):
+    """Test start_graph stops its execution if no valid credentials are provided."""
+    azure.args = MagicMock(graph_tenant_domain=None)
+    with pytest.raises(SystemExit) as err:
+        azure.start_graph()
+    assert err.value.code == 1
+    mock_logging.assert_called_once()
+
+
+@pytest.mark.parametrize('min_date, max_date, desired_date, reparse, offset, md5_hash, query', [
+    ("2022-01-15T12:00:00.000000Z", "2022-12-15T12:00:00.000000Z", "2021-12-01T12:00:00.000000Z", False, "1d", "", ""),
+    ("2021-12-01T12:00:00.000000Z", "2022-01-15T12:00:00.000000Z", "2022-12-15T12:00:00.000000Z", False, "1d", "", ""),
+    ("2021-12-01T12:00:00.000000Z", "2022-12-15T12:00:00.000000Z", "2022-01-15T12:00:00.000000Z", False, "1d", "", ""),
+    ("2000-12-01T12:00:00.000000Z", "2000-12-15T12:00:00.000000Z", "2022-01-15T12:00:00.000000Z", True, "1d", "", ""),
+])
+@patch('azure-logs.offset_to_datetime')
+@patch('azure-logs.create_new_row')
+@patch('azure-logs.orm.get_row', return_value=None)
+def test_build_graph_url(mock_get, mock_create, mock_datetime, min_date, max_date, desired_date, reparse, offset,
```

Comment:
You can also add an assert_called_with to test the logger.info call

Reply:
Done